File: ruby/6k-give-me-a-diamond/diamond.rb
def diamond(asterisks)
  # 0, negative and even numbers won't work
  return nil if asterisks < 1 or asterisks.even?

  spaces = 0
  rows = []
  while asterisks > 0
    line = Array.new
    spaces.times do line << " " end
    asterisks.times do line << "*" end
    line << "\n"

    rows << line.join
    asterisks -= 2
    spaces += 1
  end

  rows = rows[1..-1].reverse + rows
  rows.join
end

# ---------------------- #
# Tests
# ---------------------- #
def test(input, expected)
  puts "Testing with input #{input}"
  result = diamond(input)

  if result == expected
    puts " => Test passed!\n\n"
  else
    puts " => Test failed. :(\n\n"
  end
end

test(1, "*\n")
test(3, " *\n***\n *\n")
test(5, "  *\n ***\n*****\n ***\n  *\n")
test(0, nil)
test(-3, nil)
test(2, nil)

File: js/6k-give-me-a-diamond/diamond.js
// https://www.codewars.com/kata/5503013e34137eeeaa001648/
// Given a number n, return a diamond shape of asteriskes with
// n asteriskes in the middle. n-2 above and below it and so on.
// For example n = 5 should return:
//   *
//  ***
// *****
//  ***
//   *
// For negative numbers, 0 and even numbers return null.

function diamond(stars) {
    if (stars < 1 || stars % 2 === 0) {
        return null
    }

    spaces = 0
    rows = []
    rows.push(create_row(spaces, stars))
    spaces += 1
    stars -= 2

    while (stars > 0) {
        rows.unshift(create_row(spaces, stars))
        rows.push(create_row(spaces, stars))
        spaces += 1
        stars -= 2
    }

    return rows.join("\n") + "\n"
}

function create_row(spaces, stars) {
    return (" ".repeat(spaces) + "*".repeat(stars))
}

// ----------------------- //
// Tests
// ----------------------- //
function test(input, expected) {
    result = diamond(input)

    console.log("Testing with " + input + ".")
    if (result === expected) {
        console.log("Test passed!")
    } else {
        console.log("Test failed. :(")
        console.log(" => Got:      \n" + result)
        console.log("    Expected: \n" + expected)
    }
    console.log()
}

test(1, "*\n");
test(3, " *\n***\n *\n");
test(5, "  *\n ***\n*****\n ***\n  *\n");
test(2, null);
test(-3, null);
test(0, null);
